File: _includes/code/design-patterns/oop/template/template.cc
#include <fmt/core.h>
#include <string>

class Vehicle {
public:
  virtual std::string getName() { return name_; }
  virtual int getSpeed() { return speed_; }
  virtual std::string getDestination() { return destination_; }

protected:
  std::string name_ = "Vehicle";
  int speed_ = 0;
  std::string destination_ = "Unknown";
};

class Car : public Vehicle {
public:
  Car(int speed, std::string name, std::string destination)
      : speed_(speed), name_(name), destination_(destination) {}

  std::string getName() { return name_; }
  int getSpeed() { return speed_; }
  std::string getDestination() { return destination_; }

private:
  int speed_;
  std::string name_;
  std::string destination_;
};

class Ship : public Vehicle {
public:
  Ship(int speed, std::string name, std::string destination)
      : speed_(speed), name_(name), destination_(destination) {}

  std::string getName() { return name_; }
  int getSpeed() { return speed_; }
  std::string getDestination() { return destination_; }

private:
  int speed_;
  std::string name_;
  std::string destination_;
};

void printInfo(Vehicle &v) {
  fmt::print("My name is: {}, My speed is: {}, My destination is: {}\n",
             v.getName(), v.getSpeed(), v.getDestination());
}

int main() {
  auto s = Ship(5, "Boat", "Zimbabwe");
  auto c = Car(10, "Car", "Iowa");
  printInfo(s);
  printInfo(c);
}

File: _includes/code/design-patterns/oop/template/template.hs
class Vehicle a where
  name :: a -> String
  destination :: a -> String
  speed :: a -> Int

data Ship = Ship {shipName :: String, shipDestination :: String, shipSpeed :: Int } deriving (Show)
data Car = Car {carName :: String, carDestination :: String, carSpeed :: Int } deriving (Show)

instance Vehicle Ship where
  name (Ship a _ _) = a
  destination (Ship _ b _) = b
  speed (Ship _ _ c) = c

instance Vehicle Car where
  name (Car a _ _) = a
  destination (Car _ b _) = b
  speed (Car _ _ c) = c

printName :: Vehicle a => a -> IO()
printName v = putStrLn("My name is: " ++ show (name v) ++ "\n")

printSpeed :: Vehicle a => a -> IO()
printSpeed s = putStrLn("My speed is: " ++ show (speed s) ++ "\n")

printLocation :: Vehicle a => a -> IO()
printLocation s = putStrLn("My final destination is: " ++ show (destination s) ++ "\n")

main = do
  let ship = Ship "SS Shipy" "London" 10
  let car = Car "Model S" "Iowa" 55
  printSpeed car
  printSpeed ship
  printLocation car
  printLocation ship
  printName car
  printName ship
